### DISCUSSION TITLE
AG Instrument does not activate all channels

### DISCUSSION THREAD

**Discussion by Dieguitarro** [apohl79/audiogridder#1191] [Help]
Hello everyone. I had already presented this problem in the beta 15 version, apparently in this new version the problem persists.

It turns out that having one or more AG instruments inside a project does not automatically activate all channels, there is a section called "all channels" that is active by default, the problem is that the additional channels do not appear there, they appear outside of that section , so the option "all channels" cannot fulfill this function, therefore you must activate all channels manually every time you open the project

![AG Instrument](https://user-images.githubusercontent.com/123091487/228068846-add70611-5a62-4f3c-a9ae-999b9b2d64c6.jpg)

I would appreciate any solution on this.

thank you so much

**Comment by Dieguitarro** [comment]
PSD: The virtual instrument used is SUPERIOR DRUMMER 3 in the DAW cubase 12

**Comment by apohl79** [comment]
The way that menu works is:
- it presents all available buses
- a bus can have 1-many channels
- if a bus has more than 1 channel, it shows a submenu for that bus with an "All channels" item and the individual channels
- "all channels" in the submenu applies only to that bus
- at the top of the "instruments outputs" menu you have an item called "enable all channels", which applies to all buses, use this to activate all available outputs across all buses

**Reply by Dieguitarro** [reply]
Hello, thanks for the prompt response.

I understand what you are saying, I have implemented it, but AG does not save the assigned option, therefore you have to manually activate the channels every time you open projects in the DAW with similar configurations.

**Reply by apohl79** [reply]
you mean every time you add a new instance of AG, you have to go into the menu and activate all channels, right?

**Reply by Dieguitarro** [reply]
Good day.

No, every time I open an existing project with AG instrument instances I have to activate the channels manually or do the procedure you mention in the first comment, that is, AG does not save the option to "activate all channels"

**Reply by Dieguitarro** [reply]
*AG does not save the option to "enable all channels"

**Comment by Dieguitarro** [comment]
Hi, I'm sorry I didn't make myself understood.

To give a clearer and simpler example. what is happening is like using kontakt with a virtual instrument that needs multiple channels. The problem is having to activate the kontakt channels every time the daw project is opened because kontakt doesn't save the settings in the daw project. having to do that every time the daw project is opened

**Comment by Kjudkins** [comment]
How bitter sweet, I found AG seemed to be a must more efficient host for some vsti only to discover this fatal flaw on accident. 

Just to reiterate the problem: you add an instrument and it only plays the left signal of the first stereo pair. 

You remedy this by enabling all every single time, but the issue is, everything you load that project up you have to go in each single instance and re-enable the outputs by hand. The same problem occurs with SINE player, Kontakt, Spitfire player and really anything I tried. Saving the state would be one way to fix the issue, the other would be to make all outputs enabled by default. 

So close yet so far!! AG is still an interesting piece of software, would love to eventually drop VEPRO entirely.

**Reply by Dieguitarro** [reply]
Hello @Kjudkins , thank you very much for the comment.

finally someone who understands the problem, this is what I have tried to explain to @apohl79  without positive results, since in the last update the problem was not resolved.

The principle is similar to that of a host like Kontakt, imagine having to manually activate the channels every time a project is started. It's quite a headache.

I really admire the work done by @apohl79  on this plugin, it's really wonderful.... shame on this little detail

**Reply by emack999** [reply]
I do not have this issue using the Windows version, and hosting the vst in Gig Performer.  When I enable all channels for Kontakt and save, the project reloads with all channels enabled.

I think the issue is that your DAW is not saving the state properly.  Inside a DAW, AG is not responsible for remembering the state of any given session, as it cannot know one project from another.  That is the responsibility of the DAW.  If the state of a plugin isn't being remembered in one host but is recalled correctly in another, then it is the fault of the host.

**Reply by Dieguitarro** [reply]
Hello @emack999  Thanks for your comment

If what you say was correct, then any plugin similar to AG (kontakt for example) would not save the assigned configuration. I repeat, all, absolutely all plug-ins of this type save the setting, the only one that is not AG.

As seen in the image, the channels exist within AG, the problem is that they are not enabled when the project is opened, therefore it must be done manually, that is not a problem with the daw, it is with the plugin because the process is carried out internally give the plugin not externally in the Daw.

However, I would be very grateful if you could share with us the way you use AG multichannel in your DAW and that way I can verify what I am suddenly doing wrong.

**Reply by emack999** [reply]
Here is my example.
I load Kontakt into AG and look to enable all channels:
![beforesave](https://github.com/apohl79/audiogridder/assets/81657685/0d8ab0a3-134e-4a50-90d2-8b6c01dc2069)

Then I save the project, close the host.  I reopen the host, load the project I just saved, and when I look at the channels for instance of Kontakt in AG, all channels are still enabled.

![afterreload](https://github.com/apohl79/audiogridder/assets/81657685/4790b689-4d18-4cc1-b86e-49cba2dfc265)

**Reply by Dieguitarro** [reply]
Hello @emack999 . Thank you very much for the example

I do the same procedure, but it opens 2 channels by default.

What DAW do you use?

**Reply by emack999** [reply]
@Dieguitarro I use [Gig Perfomer](https://gigperformer.com/).

However, I just checked in Cubase 12 and the project saves the channels correctly there as well(using the same steps as I listed in my last post).


![cubaseAG](https://github.com/apohl79/audiogridder/assets/81657685/28b904f2-0acd-4925-b185-752173709ad0)

**Reply by emack999** [reply]
@Dieguitarro I also tested this in Cubase12 using Superior Drummer 3, as you did,  My channel settings were saved and reloaded properly.


![Cubase_SD](https://github.com/apohl79/audiogridder/assets/81657685/60758dda-0806-46f2-af9d-bc2f12d9718a)

**Reply by Dieguitarro** [reply]
![Ejemplo AGR](https://github.com/apohl79/audiogridder/assets/123091487/f12fa6d3-1eed-4159-9f5f-d738509e4113)

So I don't understand what's happening.

In my case it's as if the default plugin only enabled "all channels" found on the two stereo systems.

Is there a configuration that I am overlooking?

**Reply by Dieguitarro** [reply]
![Ejemplo 2 AGR](https://github.com/apohl79/audiogridder/assets/123091487/63f922ae-740f-4fdd-919c-63aa88eadc7b)

Does Cubase also activate the plugin channels? or how do you do it?

**Reply by emack999** [reply]
@Dieguitarro I don't know what is happening for you.
My procedure was this:

- Open Cubase > Create New project
- Add Track > Instrument > AudioGridderInst

![ScreenHunter 59](https://github.com/apohl79/audiogridder/assets/81657685/f2200549-ab1b-4fb3-9c9a-e342061b241d)

- open AG > Load Superior Drummer 3 VST >16xStereo
- In AG, Instrument Options > Enable All Channels
- Save Project > Close Cubase
- Open Cubase > load project I just saved

**Comment by emack999** [comment]
> Does Cubase also activate the plugin channels? or how do you do it?

No Cubase does not activate the plugin channels automatically.

**Reply by Dieguitarro** [reply]
@emack999   try doing the same in AG vst3. in the list of plugins it appears with 3 vertical lines III
![Ejemplo 3 AGR](https://github.com/apohl79/audiogridder/assets/123091487/1c6f80fc-c94c-4e38-90e7-53b9103b6195)

**Reply by Dieguitarro** [reply]
@apohl79 @emack999  I already found the problem.

AG VST3 plugin does not save settings.

VST2 if it saves it

@apohl79  Could you check that?

We could work with VST2 instances but the DAWs in the future will not support it and that is a big problem

**Reply by Dieguitarro** [reply]
@apohl79  The differences between the VST3 and VST2 instances of AG instrument is that when loading the VST2 plugin, by default the 64 non-active channels appear, but they do appear, in VST3 this does not happen, when loading the VST3 instance by default 1 stereo channel appears and If we activate the channels in Cubase, they appear within the plugin to activate.

For this reason I think that in the VT3 instance the default stereo channel will always appear even if it is saved with active channels.

**Comment by emack999** [comment]
> when loading the VST3 instance by default 1 stereo channel appears and If we activate the channels in Cubase, they appear within the plugin to activate.

I can confirm this behavior with the VST3 version.  When the VST3 AG plugin is loaded, only 1 Main stereo channel is visible in AG.  If you enable more channels in the DAW, then those channels appear in AG to be enabled. In the VST2 version, all 64 channels are visible in AG before doing anything in the DAW.

As a side note, I cannot change the bus layout of the AGridderInst VST3 in Gig Performer---it hangs the host.

**Comment by Kjudkins** [comment]
So is there a vsti version of AG? Or just vsti3?

I'm away from my machine but if I can atleast get stereo outs by default in
reaper I'd be back in business

**Reply by emack999** [reply]
Yes there is a vst version as well as a vst3.  They are both installed.

![ScreenHunter 61](https://github.com/apohl79/audiogridder/assets/81657685/99b8653f-a9e4-4e5c-9c0e-e387190ada75)

**Reply by Dieguitarro** [reply]
Hello @Kjudkins 

Yes, AG Instrument has two versions VSTi, VST2 and VST3, where VST2 has 64 mono channels by default to activate and save presets.

**Reply by apohl79** [reply]
Thanks guys for looking into this! Will create a bug ticket for it and see how it can be addressed.

**Reply by StephanRoemer** [reply]
Has this ever been addressed? I'm experiencing the same issue.

**Reply by hermonculous** [reply]
I guess not but I have, at least, mitigated this by editing the code and rebuilding so it defaults to all channels for the inst plugin.  Won't save selected channels, but at least enables all by default (which is what I needed for e.g. Spitfire).

Edit Plugin/Source/PluginProcessor.cpp lines 96/595

remove:
    // activate main outs per default
    // m_activeChannels.setOutputRangeActive(false);
    // for (int c = 0; c < getMainBusNumOutputChannels(); c++) {
    //     m_activeChannels.setOutputActive(c);
    // }
add:
    m_activeChannels.setOutputRangeActive(true);

**Reply by mabian69** [reply]
Do you need to only change from line 595, right? The line 96 inside the constructor can be left as it is...?

Thanks!

**Reply by mabian69** [reply]
Answering myself... I changed as mentioned, it works beautifully :)

**Reply by StephanRoemer** [reply]
What tools are needed to compile this on macOS?

**Reply by mabian69** [reply]
I managed to build it both VST2 and VST3 on Linux, including the change for channels (VST2 didn't need it anyway).
I needed to clone the git repo, both the main and dependencies, get the vst2sdk around, and follow the instructions here:

https://github.com/apohl79/audiogridder/blob/master/DEVELOP.md

Some libs were missing, and the make failed, but I found this and it helped a lot:

https://linuxmusicians.com/viewtopic.php?t=27103

On MacOS I don't really know, should be similar.

**Reply by StephanRoemer** [reply]
Thanks for the heads up! Will check out :)
